### PR TITLE
[WIP] feat: Thread titles should auto-summarize Topic

### PR DESCRIPTION
**Pull request by 0xgokuz**
Issue https://github.com/janhq/jan/issues/979

* Would like to get feedback on the architecture and structure
* If all good, then I will continue on OpenAI extensions and refactoring/cleanups
* I noticed that it's not that reliable
* `make clean` then `make dev` to test

https://github.com/janhq/jan/assets/155065525/6f0dbb68-920a-4a53-bfd9-38d9a5fdf023

### CODE REVIEW

**Comment by 0xSage** [comment]
This is cool, how are we doing thing? Pinging the local llm or a remote endpoint?

**Comment by 0xgokuz** [comment]
> This is cool, how are we doing thing? Pinging the local llm or a remote endpoint?

It would be using whatever the user is using

**Comment by 0xSage** [comment]
@imtuyethan do we have designs on this front?

I'm curious about how to balance: 
1. 20% of users want to name threads themselves
2. 80% of users want autonaming

It could be that we just deprecate 1. Or let users have a toggle/choice. What is best?

**Comment by 0xSage** [comment]
> @0xSage I think we can find the balance, similar to what ChatGPT is doing now. The autonaming kick starts ony with the first query in the thread, afterward it will not make any change. The user can change the thread title afterward.

This is a good suggestion. I'm ok with it if @imtuyethan is. We might need to think through edge cases, i.e. 
- what if user changes thread name first, then send question to llm
- race condition: what if in btw the ~5 seconds it takes for llm to generate name, user also changes thread name
etc.

**Comment by 0xgokuz** [comment]
> > @0xSage I think we can find the balance, similar to what ChatGPT is doing now. The autonaming kick starts ony with the first query in the thread, afterward it will not make any change. The user can change the thread title afterward.
> 
> This is a good suggestion. I'm ok with it if @imtuyethan is. We might need to think through edge cases, i.e.
> 
> * what if user changes thread name first, then send question to llm
> * race condition: what if in btw the ~5 seconds it takes for llm to generate name, user also changes thread name
>   etc.

I think the user's input should always overrides the "auto generate" cuz that's user's intent.

So, before we replace the title with auto generated summary from LLM, we should check if the user has renamed the Thread, if that's `true`, then just keep it.

Makes the logic simple as well. Wdyt?

File: extensions/inference-nitro-extension/src/index.ts [hunk 1]
```diff
@@ -215,13 +218,76 @@ export default class JanInferenceNitroExtension implements InferenceExtension {
     });
   }
 
+  private async onFirstPrompt(firstPrompt: MessageRequest) {
+    console.log("------ ON FIRST PROMPT NITRO ------");
+    console.log(firstPrompt);
+
+    if (firstPrompt.model.engine !== "nitro") return;
```
Comment (louis-jan) [inline extensions/inference-nitro-extension/src/index.ts:225]:
```suggestion
    if (firstPrompt.model.engine !== InferenceEngine.nitro) return;
```

File: extensions/inference-openai-extension/src/index.ts [hunk 2]
```diff
@@ -166,6 +171,11 @@ export default class JanInferenceOpenAIExtension implements InferenceExtension {
     instance.controller?.abort();
   }
 
+  private static async onFirstPrompt(firstPrompt: MessageRequest) {
+    console.log("------ ON FIRST PROMPT OPENAI ------");
+    console.log(firstPrompt);
```
Comment (louis-jan) [inline extensions/inference-openai-extension/src/index.ts:176]: Please ignore this comment if you are still working on this part, just curious if we would add the Prompt Update logic here as well.

File: web/containers/Providers/EventHandler.tsx [hunk 3]
```diff
@@ -113,13 +114,24 @@ export default function EventHandler({ children }: { children: ReactNode }) {
     }
   }
 
+  async function handleFirstPromptUpdate(message: ThreadMessage) {
+    console.log("----- HANDLE FIRST PROMPT UPDATE -----");
+    console.log(message);
+    const thread = threadsRef.current?.find((e) => e.id == message.thread_id);
+    console.log(thread);
+    if (message.content.length > 0 && message.content[0].text) {
+      thread.title = message.content[0].text.value;
```
Comment (louis-jan) [inline web/containers/Providers/EventHandler.tsx:123]: To align with product folks on how we should overwrite default title, thanks.

**Comment by 0xSage** [comment]
@0xgokuz we try to not let draft PRs dangling for long. Would you like me to close this one or are you still working on it?

**Comment by 0xgokuz** [comment]
I will just close this PR for now, and I will reopen it when it's ready @0xSage

**Pull request closed without merging**